### PR TITLE
Make Add-On Uploads More Efficient

### LINKED ISSUE #119: Updating add-on should not transfer unchanged files
https://www.widelands.org/forum/post/38363/

When updating an existing add-on, send only files that have actually changed.

I guess diffing files locally might be complicated since the "original" might not be around, but we can use checksums.
The protocol would then look like this:
- Client sends CMD_SUBMIT with full list of dirs, filenames, and checksums.
- Server sends list of files which need to be replaced:
  - Transfer all files that are not present on the server yet or whose checksum has changed.
  - If a new file has the same checksum as an existing file, we can assume that the file was moved or copied, so we don't need to request it.
  - If multiple files have the same checksum, only one copy needs to be sent.
- Client sends the files requested by the server.
- Server copies unchanged files over into the uploaded directory.

### PR DESCRIPTION
**Pull request by Noordfrees**
Fixes #119

Detects which files are already present on the server and therefore don't need to be transferred, and ensures the `addon` file is sent and validated as early as possible.

PNGs are now automatically shrunk with pngquant (if available) during uploading, for downloader-friendliness.

---

Client counterpart is at https://github.com/widelands/widelands/pull/5231

Alpha is up with this branch for testing now. (edit: not anymore)

### CODE REVIEW

**Comment by frankystone** [comment]
Do i need a specific widelands branch for testing?

Uploading my minimalistic theme with the trident-fixes branch the pngs, which i didn't change, get all uploaded. So the upload time is not reduced. pngquant is not installed here.

**Comment by Noordfrees** [comment]
Bug: In the maintainer e-mail, `Old version` is given as `null`

**Comment by frankystone** [comment]
Uploaded minimalstic_theme and this took just a few seconds. Without pngquant.

**Comment by Noordfrees** [comment]
> Uploaded minimalstic_theme and this took just a few seconds. 

Good :)

> Without pngquant.

PNGQuant is run on the server, not on the client. We can hardly expect every player to install it ;) I installed it in the containers; however it is only run on newly uploaded files, so it won't have had an effect yet since no PNG files were uploaded.

**Comment by frankystone** [comment]
So a test for pngquant is also needed? Can you delete the minimalistic_theme on alpha? I don't want to create/add new pngs for each test...

It would be nice to have a visual representation to which server one is connected. Starting widelands with the `--verbose`-flag says "Resolved network name 'widelands.org:7377' to ....", but this sentence is not very helpful for people who do not know what it means. Is it possible to add the addon_server to the title of the window? E.g. if one is connected with the alpha-server the window titel is "Add-On manager: Connected to alpha". If connected with the productive server the window title can stay as it is right now.
This is also important because the command line argument is stored in the widelands config and get applied automatically on next run. This can be confusing.

**Comment by frankystone** [comment]
Got two(!) "AddOn deleted" emails for the deletion of minimalistic_theme.

I am AFK for some while now.

**Comment by Noordfrees** [comment]
> Got two(!) "AddOn deleted" emails for the deletion of minimalistic_theme.

That's fine, because I somehow registered *two* of your alpha accounts as uploaders for each of your add-ons, so each account received one message

> It would be nice to have a visual representation to which server one is connected.

+1
Could be done by letting each server define a name in its config file and advertise this name in the initial contact, and the Widelands UI would then display it (e.g. "Alpha" or "Local test environment", or "" for the official server).

**Comment by frankystone** [comment]
Uploaded again, last very long as expected. Then i deleted the addon in widelands and downloaded again which takes less than a minute. :+1:

**Comment by Noordfrees** [comment]
Please submit another update for the minimalistic-theme add-on to alpha so I can check whether the e-mail is now formatted correctly. If that works this should be good to go :)

**Comment by frankystone** [comment]
Done.

**Comment by Noordfrees** [comment]
Thanks :) 
E-Mail worked correctly.

---

I'm going to change alpha to run #128 now…

File: wl/server/Command.java [hunk 1]
```diff
@@ -229,12 +229,16 @@
 	 *   <li> Integer string denoting number <var>D</var> of directories, <code>\n</code>
```
[the remark this answers is not in the document]
Comment (Noordfrees) [reply wl/server/Command.java:229]: This is sort of standard practice when writing javadocs, see e.g. https://stackoverflow.com/q/33787753
I never dare omit any closing tags when writing real HTML files, but in code comments it's more readable IMHO to signify only where the next paragraph or item begins and leave out the optional tags.

**Comment by hessenfarmer** [comment]
@frankystone 
Anything to change? Or could you approve this PR.

**Review by frankystone — APPROVED** [review]
Sorry for the late response.

**Comment by Noordfrees** [comment]
Thanks :) 
Deployed.

**Pull request merged**